src/components/application_manager/src/commands/button_notification_to_mobile.cc

```diff
@@ -0,0 +1,209 @@
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following
+ * disclaimer in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ * Neither the name of the Ford Motor Company nor the names of its contributors
+ * may be used to endorse or promote products derived from this software
+ * without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "application_manager/commands/button_notification_to_mobile.h"
+namespace application_manager {
+
+namespace commands {
+namespace app_mngr = application_manager;
+
+SDL_CREATE_LOG_VARIABLE("ButtonNotificationToMobile")
+
+ButtonNotificationToMobile::ButtonNotificationToMobile(
+    const app_mngr::commands::MessageSharedPtr& message,
+    app_mngr::ApplicationManager& application_manager,
+    app_mngr::rpc_service::RPCService& rpc_service,
+    app_mngr::HMICapabilities& hmi_capabilities,
+    policy::PolicyHandlerInterface& policy_handler)
+    : CommandNotificationImpl(message,
+                              application_manager,
+                              rpc_service,
+                              hmi_capabilities,
+                              policy_handler) {}
+
+ButtonNotificationToMobile::~ButtonNotificationToMobile() {}
+
+bool ButtonNotificationToMobile::DoesAppIDExist() const {
+  SDL_LOG_AUTO_TRACE();
+  using namespace application_manager::strings;
+  return (*message_)[msg_params].keyExists(strings::app_id);
+}
+
+void ButtonNotificationToMobile::HandleCustomButton(
+    app_mngr::ApplicationSharedPtr app) {
+  SDL_LOG_AUTO_TRACE();
+  using namespace application_manager::strings;
+  // app_id is mandatory for CUSTOM_BUTTON notification
+  if (!DoesAppIDExist()) {
+    SDL_LOG_ERROR("CUSTOM_BUTTON mobile notification without app_id.");
+    return;
+  }
+
+  // custom_button_id is mandatory for CUSTOM_BUTTON notification
+  if (!(*message_)[msg_params].keyExists(hmi_response::custom_button_id)) {
+    SDL_LOG_ERROR(
+        "CUSTOM_BUTTON mobile notification without custom_button_id.");
+    return;
+  }
+
+  if (!app) {
```

Comment:
it's better to move this check to the beginning of function as it is input parameter

Reply:
changed in a65c63788fef72c92bd702270a97f27a8f34d4f4